### PR TITLE
Spike orm bucket split

### PR DESCRIPTION
**Pull request by alpe**
Heavy ORM bucket redesign
The idea was to introduce chainable buckets for easy extension. Supports #752

Please check for examples of the last modified updating bucket:
* `aswap` bucket + handler and 
* `distribution` bucket + handler first.
* The interfaces in `orm/bucket.go`....

### CODE REVIEW

**Comment by iov-bot-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/iov-one/weave/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: x/aswap/model.go [hunk 1]
```diff
@@ -72,24 +72,25 @@ func AsSwap(obj orm.Object) *Swap {
...
 
 // NewBucket initializes a Bucket with default name
 //
-// inherit Get and Save from orm.Bucket
+// inherit Get and Save from orm.BaseBucket
 // add Create
 func NewBucket() Bucket {
-	bucket := migration.NewBucket("aswap", BucketName,
-		orm.NewSimpleObj(nil, &Swap{})).
+	bucket := orm.NewBucketBuilder(BucketName, orm.NewSimpleObj(nil, &Swap{})).
```
Comment (ruseinov) [inline x/aswap/model.go]: Yeah, builder seems like a nice way to do this. Reminds me a lot of how you assemble a complex object in Java, but that's what we have - a pretty complex struct with a bunch of options.

I can't help but wonder if we can include `WithMigration` in the builder as well. 

One other thing we could do is include all possible bucket modifications into the builder and then try to convert resulting `built` bucket with a helper, there we can check if all the conditions are satisfied. But maybe this just complicates things even more.

**Comment by iov-bot-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/iov-one/weave/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by ethanfrey** [comment]
I started reviewing this, but wasn't sure where to start (maybe I'm just too tired now).

I see the new interfaces defined in orm. Is that the key?
For spikes, I prefer much smaller changesets so I can see the demo of the idea, doesn't really matter if only two packages compile, the ones you modify.

**Comment by ethanfrey** [comment]
Seems a bit odd to have this generic Bucket abstraction, then we must remember to call `orm.MustBindCtxHeight(h.bucket, ctx)`.

If it requires more arguments, shouldn't we make that explicit?
Or do we Bind for all Buckets?

I know it makes changes easier from current state, but it seems like a magic, optional step that is really required. Adding an argument to a call makes the compiler red. Slipping this in must be entirely optional... but it is not.

**Comment by husio** [comment]
I got a private tour through the changes from Alex yesterday and we discussed it afterward. I will repeat here my comments as not everyone could here me :wink:

I think implementing this functionality for all buckets is a YUGE task. This change results in 7 bucket interfaces.
I thought the plan was to migrate code to use `ModelBucket` implementation and delete all others. https://github.com/iov-one/weave/pull/615#pullrequestreview-237443106


I like the `Bind` method. I think the details of how it should work is a big decision (I will come back to this later). [This implementation](https://github.com/iov-one/weave/pull/615#pullrequestreview-237443106) is modyfing an existing entity which I think is dangerous. You bind it to height 3 in one handler and until it is bind to different height, all users use height 3 bucket. This can cause hard to find bugs as we share referernce to the same bucket between handlers.
I think [`BindHeight`](https://github.com/iov-one/weave/pull/775/files#diff-f5a56dc0b59391148c8fe47edd3862f5R44) trying to find down the stack an object it could mutate could be replaced with an explicit wrapper.

I believe there are two ways a `Bind` of a bucket could work.
1. A bucket provides read only functionality. When it it bound to a block height, it exlose additional functionality that allows to modify the data.
2. A bucket provides no functionality until it is bound to a block height. This is a subset of the first solution.

In both cases, compiler can assure that no logic violation is possible. I have implementad [a proof of concept](https://github.com/iov-one/weave/commit/c3c1884109e35609be85fcce6f03f35350f7bf85#diff-61d5e52e69a66062832b227256f02752R11) for the second choice and ported `x/paychan` extension. It could be easily updated to implement the first idea as well.


The last point is the final complexity of the solution we are building. The current [`orm.Bucket`](https://godoc.org/github.com/iov-one/weave/orm#Bucket) has a broad interface. We have learned from using it and created a new, much smaller [`orm.ModelBucket`](https://godoc.org/github.com/iov-one/weave/orm#ModelBucket) that provides the very same functionality.
I believe that we should strive to provide small and reusable interfaces. When a new person is introduced to weave codebase, it might be overwhelming to have to choose from 7+ bucket implementatinos and plenty of options.
I am not saying we should limit our functionality. We do need migrations, automatic ID generation is nice, we want to store block height. What I am saying is that we should be cacious with broadening the API.
I also think that explicit and verbose is often better than magical. It is more typing that you do once. It is much easier read that many people will thank us for later. :older_man:

File: x/aswap/model.go (hunk 1 above)
Comment (alpe) [reply x/aswap/model.go]: `WithMigration` causes cyclomatic dependencies unfortunately.

**Comment by alpe** [comment]
Closing as the discussion has evolved to a new design approach.

**Comment by ethanfrey** [comment]
Nice comments @husio 

Those echoed my feelings as well.

I very much like the idea of the BucketBuilder (so you cannot add indexes to a bucket outside the constructor). And some of the interface methods are very well names (create, update, delete - wonderful). However, the general design adds more complexity.

I would like something like:

* `orm.ModelBucket` - does all basic stuff but migrations (as out of orm)
* `migrations.ModelBucket` - same interface as ModelBucket but add migrations checks
* If needed some transparent enhancements can wrap these - but MUST keep same interface

`VersionedModelBucket` - can accept any `ModelBucket` interface and expose a *different* API with versions front and center. this is special-purpose and is not intended to be wrapped by anything else.

**Pull request closed without merging**